clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +243,461 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+        }
+    }
+
+    /**
+     * Perform a topic metadata request to discover topic names for the given topic ids.
+     *
+     * @param topicsRequiringMetadata List of topic Uuid for which topic names are needed.
+     * @param resolvedTopicPartitions List of TopicPartitions for the topics with known names.
+     *                                This list will be extended when the missing topic names are
+     *                                received in metadata.
+     *
+     * @return Future that will complete when topic names are received for all
+     * topicsRequiringMetadata. It will fail if a metadata response is received but does not
+     * include all the topics that were requested.
+     */
+    private CompletableFuture<SortedSet<TopicPartition>> resolveTopicNamesForTopicIds(
+            List<Uuid> topicsRequiringMetadata,
+            SortedSet<TopicPartition> resolvedTopicPartitions) {
+        CompletableFuture<SortedSet<TopicPartition>> result = new CompletableFuture<>();
+        log.debug("Topic IDs {} received in assignment were not found in metadata. " +
+                "Requesting metadata to resolve topic names and proceed with the " +
+                "reconciliation.", topicsRequiringMetadata);
+        // TODO: request metadata only for the topics that require it. Passing empty list to
+        //  retrieve it for all topics until the TopicMetadataRequestManager supports a list
+        //  of topics.
+        CompletableFuture<Map<Topic, List<PartitionInfo>>> metadataResult = metadataRequestManager.requestTopicMetadata(Optional.empty());
+        metadataResult.whenComplete((topicNameAndPartitionInfo, error) -> {
+            if (error != null) {
+                // Metadata request to get topic names failed. The TopicMetadataManager
+                // handles retries on retriable errors, so at this point we consider this a
+                // fatal error.
+                log.error("Metadata request for topic IDs {} received in assignment failed.",
+                        topicsRequiringMetadata, error);
+                result.completeExceptionally(new KafkaException("Failed to get metadata for " +
+                        "topic IDs received in target assignment.", error));
+            } else {
+                topicNameAndPartitionInfo.forEach((topic, partitionInfoList) -> {
+                    if (topicsRequiringMetadata.contains(topic.topicId())) {
+                        partitionInfoList.forEach(partitionInfo ->
+                                resolvedTopicPartitions.add(new TopicPartition(topic.topicName(), partitionInfo.partition())));
+                        topicsRequiringMetadata.remove(topic.topicId());
+                    }
+                });
+                if (topicsRequiringMetadata.isEmpty()) {
+                    result.complete(resolvedTopicPartitions);
+                } else {
+                    // TODO: check if this could happen. If so, we probably need to retry the
```

Comment:
I think that theoretically it could and the action you've proposed is correct.

Reply:
Thanks for confirming @AndrewJSchofield . After the change to integrate this with the centralized metadata object and cache, we do achieve this behaviour (we keep retrying until all assigned topic IDs are found in metadata)